File: merge_sort.cpp
#include <bits/stdc++.h>
using namespace std;

void merge(int arr[], int left, int mid, int right)
{
	int subArrayOne = mid - left + 1;
	int subArrayTwo = right - mid;

	int* leftArray = new int[subArrayOne];
	int* rightArray = new int[subArrayTwo];

	for(int i = 0; i < subArrayOne; i++)
		leftArray[i] = arr[left + i];
	for(int j = 0; j < subArrayTwo; j++)
		rightArray[j] = arr[mid + 1 + j];

	int indexOfSubArrayOne = 0,indexOfSubArrayTwo = 0;
	int indexOfMergedArray = left;

	while (indexOfSubArrayOne < subArrayOne && indexOfSubArrayTwo < subArrayTwo) 
    {
		if (leftArray[indexOfSubArrayOne] <= rightArray[indexOfSubArrayTwo]) 
        {
			arr[indexOfMergedArray] = leftArray[indexOfSubArrayOne];
			indexOfSubArrayOne++;
		}
		else 
        {
			arr[indexOfMergedArray] = rightArray[indexOfSubArrayTwo];
			indexOfSubArrayTwo++;
		}
		indexOfMergedArray++;
	}

	while (indexOfSubArrayOne < subArrayOne) 
    {
		arr[indexOfMergedArray] = leftArray[indexOfSubArrayOne];
		indexOfSubArrayOne++;
		indexOfMergedArray++;
	}
	while (indexOfSubArrayTwo < subArrayTwo) 
    {
		arr[indexOfMergedArray] = rightArray[indexOfSubArrayTwo];
		indexOfSubArrayTwo++;
		indexOfMergedArray++;
	}
}

void mergeSort(int arr[], int begin, int end)
{
	if (begin >= end)
		return;

	auto mid = begin + (end - begin) / 2;
	mergeSort(arr, begin, mid);
	mergeSort(arr, mid + 1, end);
	merge(arr, begin, mid, end);
}

void printArray(int a[], int size)
{
	for(int i = 0;i<size; i++)
		cout << a[i] << " ";
}

int main()
{
    int n;
    cout << "Enter size of array: ";
    cin >> n;
    int arr[n];
    cout << "Enter elemments of array: ";
    for(int i=0;i<n;i++)
        cin >> arr[i];
	mergeSort(arr,0,n-1);
	cout << "Sorted array is: ";
	printArray(arr,n);
	return 0;
}

File: MergeSort.java
public class Mergesort {

    public static void printarr(int arr[]) {

        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }

        System.out.println();
    }

    // DIVIDE
    public static void sort(int arr[], int si, int ei) {

        // Base condition
        if (si >= ei) {
            return;
        }

        int mid = si + (ei - si) / 2;

        // left side
        sort(arr, si, mid);

        // right side
        sort(arr, mid + 1, ei);

        // merging both the left and right side
        merge(arr, si, ei, mid);
    }

    // CONQUER
    // Function for merging the left and right array
    public static void merge(int arr[], int si, int ei, int mid) {

        // firstly, we will create a temporary array to merge the left and right side of
        // the array
        int temp[] = new int[ei - si + 1]; // size of the new array.

        // Now we will make the iterator for both the parts
        int i = si; // left part
        int j = mid + 1; // right part
        int k = 0; // temp array

        while (i <= mid && j <= ei) {
            if (arr[i] < arr[j]) {
                temp[k] = arr[i];
                i++;
            } else {
                temp[k] = arr[j];
                j++;
            }
            k++;
        }
        while (i <= mid) {
            temp[k] = arr[i];
            i++;
            k++;
        }
        while (j <= ei) {
            temp[k] = arr[j];
            j++;
            k++;
        }

        // copy temporary to my original array
        for (k = 0, i = si; k < temp.length; k++, i++) {
            arr[i] = temp[k];
        }
    }

    public static void main(String args[]) {

        int arr[] = { 2, 5, 8, 4, 9, 3 };

        sort(arr, 0, arr.length - 1); // Remember: ei is arr.length-1

        printarr(arr);
    }
}
// We use Depth first search to implement Merge sort